File: ics-runtime/ics_runtime/session_backends/sqlite_backend.py
"""SQLite session backend — file-based persistence without external deps."""

from __future__ import annotations

import json
import sqlite3
import threading
from datetime import datetime, timezone
from pathlib import Path

from ics_runtime.session_backends.base import SessionBackend, SessionData


_DDL = """
CREATE TABLE IF NOT EXISTS sessions (
    session_id TEXT PRIMARY KEY,
    data       TEXT NOT NULL,
    created_at TEXT NOT NULL,
    updated_at TEXT NOT NULL
)
"""


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _serialize(data: SessionData) -> str:
    return json.dumps({
        "session_id": data.session_id,
        "entries": data.entries,
        "context": data.context,
        "cleared": data.cleared,
        "created_at": data.created_at.isoformat(),
        "last_active": data.last_active.isoformat(),
        "turn_count": data.turn_count,
    })


def _deserialize(raw: str) -> SessionData:
    d = json.loads(raw)
    return SessionData(
        session_id=d["session_id"],
        entries=d.get("entries", []),
        context=d.get("context", {}),
        cleared=d.get("cleared", False),
        created_at=datetime.fromisoformat(d["created_at"]),
        last_active=datetime.fromisoformat(d["last_active"]),
        turn_count=d.get("turn_count", 0),
    )


class SQLiteBackend(SessionBackend):
    """Stores sessions in a SQLite database file.

    Suitable for single-process applications that need session persistence
    across restarts without running a Redis server.

    Args:
        db_path: Path to the SQLite file.  Use ``":memory:"`` for tests.
    """

    def __init__(self, db_path: str | Path = "ics_sessions.db") -> None:
        self._db_path = str(db_path)
        self._lock = threading.Lock()
        # For :memory: databases keep a single persistent connection; each new
        # connect() call would create a fresh empty DB.
        if self._db_path == ":memory:":
            self._persistent_conn: sqlite3.Connection | None = sqlite3.connect(
                ":memory:", check_same_thread=False
            )
            self._persistent_conn.row_factory = sqlite3.Row
        else:
            self._persistent_conn = None
        self._init_db()

    def _connect(self) -> sqlite3.Connection:
        if self._persistent_conn is not None:
            return self._persistent_conn
        conn = sqlite3.connect(self._db_path, check_same_thread=False)
        conn.row_factory = sqlite3.Row
        return conn

    def _init_db(self) -> None:
        conn = self._connect()
        with self._lock:
            conn.execute(_DDL)
            conn.commit()

    def load(self, session_id: str) -> SessionData | None:
        conn = self._connect()
        with self._lock:
            row = conn.execute(
                "SELECT data FROM sessions WHERE session_id = ?", (session_id,)
            ).fetchone()
        if row is None:
            return None
        return _deserialize(row["data"])

    def save(self, session_id: str, data: SessionData) -> None:
        now = _now()
        serialized = _serialize(data)
        conn = self._connect()
        with self._lock:
            conn.execute(
                """
                INSERT INTO sessions (session_id, data, created_at, updated_at)
                VALUES (?, ?, ?, ?)
                ON CONFLICT(session_id) DO UPDATE SET data = excluded.data,
                    updated_at = excluded.updated_at
                """,
                (session_id, serialized, now, now),
            )
            conn.commit()

    def delete(self, session_id: str) -> None:
        conn = self._connect()
        with self._lock:
            conn.execute("DELETE FROM sessions WHERE session_id = ?", (session_id,))
            conn.commit()

    def exists(self, session_id: str) -> bool:
        conn = self._connect()
        with self._lock:
            row = conn.execute(
                "SELECT 1 FROM sessions WHERE session_id = ?", (session_id,)
            ).fetchone()
        return row is not None

File: ics-runtime/java/src/main/java/io/ics/runtime/backends/SQLiteBackend.java
package io.ics.runtime.backends;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.sql.*;
import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * SQLite-backed session storage.
 *
 * <p>Pass {@code ":memory:"} as the path for an in-process SQLite database
 * (uses a persistent connection so tables survive between calls, matching the
 * Python implementation's fix for the same problem).
 *
 * <p>Uses the JDBC SQLite driver bundled with the Anthropic/OpenAI SDKs via
 * the transitive dependency {@code org.xerial:sqlite-jdbc}.  Add it explicitly
 * if needed:
 * <pre>{@code
 * <dependency>
 *   <groupId>org.xerial</groupId>
 *   <artifactId>sqlite-jdbc</artifactId>
 *   <version>3.46.0.0</version>
 * </dependency>
 * }</pre>
 */
public final class SQLiteBackend implements SessionBackend {

    private static final ObjectMapper JSON = new ObjectMapper();

    private final String path;
    private final Connection persistentConn;   // non-null only for ":memory:"

    public SQLiteBackend(String path) {
        this.path = path;
        if (":memory:".equals(path)) {
            try {
                persistentConn = DriverManager.getConnection("jdbc:sqlite::memory:");
                initDb(persistentConn);
            } catch (SQLException e) {
                throw new RuntimeException("Failed to initialise in-memory SQLite", e);
            }
        } else {
            persistentConn = null;
            try (Connection c = connect()) {
                initDb(c);
            } catch (SQLException e) {
                throw new RuntimeException("Failed to initialise SQLite at " + path, e);
            }
        }
    }

    // ── SessionBackend ────────────────────────────────────────────────────────

    @Override
    public SessionData load(String sessionId) {
        String sql = "SELECT entries, context, cleared, created_at, last_active, turn_count " +
                     "FROM sessions WHERE session_id = ?";
        try (Connection c = connect();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, sessionId);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) return null;

                SessionData data = new SessionData(sessionId);
                String entriesJson  = rs.getString("entries");
                String contextJson  = rs.getString("context");
                data.setCleared(rs.getBoolean("cleared"));
                data.setLastActive(Instant.parse(rs.getString("last_active")));
                // Restore turn_count via reflection-free approach: increment externally
                int tc = rs.getInt("turn_count");
                for (int i = 0; i < tc; i++) data.incrementTurnCount();

                if (entriesJson != null) {
                    List<String> entries = JSON.readValue(entriesJson, new TypeReference<>(){});
                    entries.forEach(data::addEntry);
                }
                if (contextJson != null) {
                    Map<String, Object> ctx = JSON.readValue(contextJson, new TypeReference<>(){});
                    data.getContext().putAll(ctx);
                }
                return data;
            }
        } catch (Exception e) {
            throw new RuntimeException("SQLite load failed for " + sessionId, e);
        }
    }

    @Override
    public void save(String sessionId, SessionData data) {
        String sql = "INSERT INTO sessions (session_id, entries, context, cleared, created_at, last_active, turn_count) " +
                     "VALUES (?, ?, ?, ?, ?, ?, ?) " +
                     "ON CONFLICT(session_id) DO UPDATE SET " +
                     "entries=excluded.entries, context=excluded.context, " +
                     "cleared=excluded.cleared, last_active=excluded.last_active, " +
                     "turn_count=excluded.turn_count";
        try (Connection c = connect();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, sessionId);
            ps.setString(2, JSON.writeValueAsString(data.getEntries()));
            ps.setString(3, JSON.writeValueAsString(data.getContext()));
            ps.setBoolean(4, data.isCleared());
            ps.setString(5, data.getCreatedAt().toString());
            ps.setString(6, Instant.now().toString());
            ps.setInt(7, data.getTurnCount());
            ps.executeUpdate();
        } catch (Exception e) {
            throw new RuntimeException("SQLite save failed for " + sessionId, e);
        }
    }

    @Override
    public void delete(String sessionId) {
        try (Connection c = connect();
             PreparedStatement ps = c.prepareStatement("DELETE FROM sessions WHERE session_id = ?")) {
            ps.setString(1, sessionId);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException("SQLite delete failed for " + sessionId, e);
        }
    }

    @Override
    public boolean exists(String sessionId) {
        try (Connection c = connect();
             PreparedStatement ps = c.prepareStatement(
                     "SELECT 1 FROM sessions WHERE session_id = ?")) {
            ps.setString(1, sessionId);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next();
            }
        } catch (SQLException e) {
            return false;
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private Connection connect() throws SQLException {
        if (persistentConn != null) return persistentConn;
        return DriverManager.getConnection("jdbc:sqlite:" + path);
    }

    private static void initDb(Connection c) throws SQLException {
        try (Statement st = c.createStatement()) {
            st.execute("""
                CREATE TABLE IF NOT EXISTS sessions (
                    session_id  TEXT PRIMARY KEY,
                    entries     TEXT NOT NULL DEFAULT '[]',
                    context     TEXT NOT NULL DEFAULT '{}',
                    cleared     INTEGER NOT NULL DEFAULT 0,
                    created_at  TEXT NOT NULL,
                    last_active TEXT NOT NULL,
                    turn_count  INTEGER NOT NULL DEFAULT 0
                )""");
        }
    }
}
